File: packages/shared/src/geometry.ts
export interface Rect { x: number; y: number; width: number; height: number }
export type Point = { x: number; y: number; p?: number };

export const rect = (x: number, y: number, width: number, height: number): Rect => ({ x, y, width, height });

export function rectRight(r: Rect): number { return r.x + r.width; }
export function rectBottom(r: Rect): number { return r.y + r.height; }

export function rectCenter(r: Rect): { x: number; y: number } {
  return { x: r.x + r.width / 2, y: r.y + r.height / 2 };
}

export function expandRect(r: Rect, by: number): Rect {
  return { x: r.x - by, y: r.y - by, width: r.width + by * 2, height: r.height + by * 2 };
}

export function rectsIntersect(a: Rect, b: Rect): boolean {
  return a.x < rectRight(b) && b.x < rectRight(a) && a.y < rectBottom(b) && b.y < rectBottom(a);
}

export function rectContains(outer: Rect, inner: Rect): boolean {
  return (
    inner.x >= outer.x &&
    inner.y >= outer.y &&
    rectRight(inner) <= rectRight(outer) &&
    rectBottom(inner) <= rectBottom(outer)
  );
}

export function unionRect(a: Rect, b: Rect): Rect {
  const x = Math.min(a.x, b.x);
  const y = Math.min(a.y, b.y);
  return { x, y, width: Math.max(rectRight(a), rectRight(b)) - x, height: Math.max(rectBottom(a), rectBottom(b)) - y };
}

export function unionRects(rects: readonly Rect[]): Rect | null {
  if (rects.length === 0) return null;
  let acc = rects[0]!;
  for (let i = 1; i < rects.length; i++) acc = unionRect(acc, rects[i]!);
  return acc;
}

export function intersectRect(a: Rect, b: Rect): Rect | null {
  const x = Math.max(a.x, b.x);
  const y = Math.max(a.y, b.y);
  const r = Math.min(rectRight(a), rectRight(b));
  const bo = Math.min(rectBottom(a), rectBottom(b));
  if (r <= x || bo <= y) return null;
  return { x, y, width: r - x, height: bo - y };
}

/** Clamp a rect so it stays inside [0,bounds]^2 without changing its size (size is clipped if larger). */
export function clampRectInside(r: Rect, boundsWidth: number, boundsHeight: number): Rect {
  const width = Math.min(r.width, boundsWidth);
  const height = Math.min(r.height, boundsHeight);
  const x = Math.min(Math.max(r.x, 0), boundsWidth - width);
  const y = Math.min(Math.max(r.y, 0), boundsHeight - height);
  return { x, y, width, height };
}

export function roundRectValues(r: Rect): Rect {
  const x = Math.floor(r.x);
  const y = Math.floor(r.y);
  return { x, y, width: Math.ceil(rectRight(r)) - x, height: Math.ceil(rectBottom(r)) - y };
}

/** Bounding box of a stroke, padded by half its width (plus 1 for antialiasing). */
export function strokeBBox(points: readonly Point[], width: number): Rect {
  if (points.length === 0) return { x: 0, y: 0, width: 0, height: 0 };
  let minX = Infinity, minY = Infinity, maxX = -Infinity, maxY = -Infinity;
  for (const pt of points) {
    if (pt.x < minX) minX = pt.x;
    if (pt.y < minY) minY = pt.y;
    if (pt.x > maxX) maxX = pt.x;
    if (pt.y > maxY) maxY = pt.y;
  }
  const pad = width / 2 + 1;
  return { x: minX - pad, y: minY - pad, width: maxX - minX + pad * 2, height: maxY - minY + pad * 2 };
}

File: apps/brushjam/src/brushjam/geometry.py
"""Mirror of packages/shared/src/geometry.ts."""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Iterable, List, Optional, Sequence, TypedDict


class Rect(TypedDict):
    x: float
    y: float
    width: float
    height: float


class Point(TypedDict, total=False):
    x: float
    y: float
    p: float


def rect(x: float, y: float, width: float, height: float) -> Rect:
    return {"x": x, "y": y, "width": width, "height": height}


def rect_right(r: Rect) -> float:
    return r["x"] + r["width"]


def rect_bottom(r: Rect) -> float:
    return r["y"] + r["height"]


def translate_rect(r: Rect, dx: float, dy: float) -> Rect:
    return {"x": r["x"] + dx, "y": r["y"] + dy, "width": r["width"], "height": r["height"]}


def expand_rect(r: Rect, by: float) -> Rect:
    return {"x": r["x"] - by, "y": r["y"] - by, "width": r["width"] + by * 2, "height": r["height"] + by * 2}


def rects_intersect(a: Rect, b: Rect) -> bool:
    return (
        a["x"] < rect_right(b)
        and b["x"] < rect_right(a)
        and a["y"] < rect_bottom(b)
        and b["y"] < rect_bottom(a)
    )


def union_rect(a: Rect, b: Rect) -> Rect:
    x = min(a["x"], b["x"])
    y = min(a["y"], b["y"])
    return {
        "x": x,
        "y": y,
        "width": max(rect_right(a), rect_right(b)) - x,
        "height": max(rect_bottom(a), rect_bottom(b)) - y,
    }


def union_rects(rects: Sequence[Rect]) -> Optional[Rect]:
    if not rects:
        return None
    acc = rects[0]
    for r in rects[1:]:
        acc = union_rect(acc, r)
    return acc


def intersect_rect(a: Rect, b: Rect) -> Optional[Rect]:
    x = max(a["x"], b["x"])
    y = max(a["y"], b["y"])
    r = min(rect_right(a), rect_right(b))
    bo = min(rect_bottom(a), rect_bottom(b))
    if r <= x or bo <= y:
        return None
    return {"x": x, "y": y, "width": r - x, "height": bo - y}


def clamp_rect_inside(r: Rect, bounds_width: float, bounds_height: float) -> Rect:
    width = min(r["width"], bounds_width)
    height = min(r["height"], bounds_height)
    x = min(max(r["x"], 0), bounds_width - width)
    y = min(max(r["y"], 0), bounds_height - height)
    return {"x": x, "y": y, "width": width, "height": height}


def round_rect_values(r: Rect) -> Rect:
    x = math.floor(r["x"])
    y = math.floor(r["y"])
    return {"x": x, "y": y, "width": math.ceil(rect_right(r)) - x, "height": math.ceil(rect_bottom(r)) - y}


def stroke_bbox(points: Sequence[Point], width: float) -> Rect:
    """Bounding box of a stroke, padded by half its width (plus 1 for AA)."""
    if not points:
        return {"x": 0, "y": 0, "width": 0, "height": 0}
    min_x = min(p["x"] for p in points)
    min_y = min(p["y"] for p in points)
    max_x = max(p["x"] for p in points)
    max_y = max(p["y"] for p in points)
    pad = width / 2 + 1
    return {
        "x": min_x - pad,
        "y": min_y - pad,
        "width": max_x - min_x + pad * 2,
        "height": max_y - min_y + pad * 2,
    }


def subtract_rect(a: Rect, b: Rect) -> List[Rect]:
    """`a` minus `b`, as up to four rects."""
    overlap = intersect_rect(a, b)
    if not overlap:
        return [a]
    out: List[Rect] = []
    if overlap["y"] > a["y"]:
        out.append({"x": a["x"], "y": a["y"], "width": a["width"], "height": overlap["y"] - a["y"]})
    if rect_bottom(overlap) < rect_bottom(a):
        out.append(
            {
                "x": a["x"],
                "y": rect_bottom(overlap),
                "width": a["width"],
                "height": rect_bottom(a) - rect_bottom(overlap),
            }
        )
    if overlap["x"] > a["x"]:
        out.append({"x": a["x"], "y": overlap["y"], "width": overlap["x"] - a["x"], "height": overlap["height"]})
    if rect_right(overlap) < rect_right(a):
        out.append(
            {
                "x": rect_right(overlap),
                "y": overlap["y"],
                "width": rect_right(a) - rect_right(overlap),
                "height": overlap["height"],
            }
        )
    return out
